### PR TITLE
#1797 Use EnumSet.noneOf when creating a new instance of EnumSet

### LINKED ISSUE #1797: [REGRESSION] Make MapStruct generate compilable mapping code for attributes of type EnumSet
See #732, does not work anymore

### PR DESCRIPTION
**Pull request by filiphr**
Fixes #1797

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/IterableCreation.java [hunk 1]
```diff
@@ -69,6 +71,20 @@ public boolean isLoadFactorAdjustment() {
...
+        }
         return types;
     }
+
+    public Type getEnumSetElementType() {
+        return first( getResultType().determineTypeArguments( Iterable.class ) );
+    }
+
+    public boolean isEnumSet() {
+        return "java.util.EnumSet".equals( resultType.getFullyQualifiedName() );
```
Comment (chris922) [inline processor/src/main/java/org/mapstruct/ap/internal/model/IterableCreation.java]: What about this?

```suggestion
        return EnumSet.class.getName().equals( resultType.getFullyQualifiedName() );
```

File: processor/src/main/resources/org/mapstruct/ap/internal/model/IterableCreation.ftl [hunk 2]
```diff
@@ -9,6 +9,8 @@
 <@compress single_line=true>
     <#if factoryMethod??>
         <@includeModel object=factoryMethod targetType=resultType/>
+    <#elseif enumSet>
+        EnumSet.noneOf( <@includeModel object=enumSetElementType raw=true/>.class )
```
Comment (chris922) [inline processor/src/main/resources/org/mapstruct/ap/internal/model/IterableCreation.ftl]: Is there somehow a way to do this more generic and don't hard-code the enum-set "instantiation" here in the template? Like the previous branch is there for all kind of factory methods.. or just keep it simple as we don't have any other use-case up to now?

Comment (filiphr) [reply processor/src/main/resources/org/mapstruct/ap/internal/model/IterableCreation.ftl]: The reason why I did it like that was that it was already done in a similar way in [`ExistingInstanceSetterWrapperForCollectionsAndMaps.ftl`](https://github.com/mapstruct/mapstruct/blob/871353fccbf8ed05cfc7dee055f99eb81b0e69e0/processor/src/main/resources/org/mapstruct/ap/internal/model/assignment/ExistingInstanceSetterWrapperForCollectionsAndMaps.ftl#L39) and [`SetterWrapperForCollectionsAndMapsWithNullCheck.ftl`](https://github.com/mapstruct/mapstruct/blob/871353fccbf8ed05cfc7dee055f99eb81b0e69e0/processor/src/main/resources/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMapsWithNullCheck.ftl#L27).

I was thinking whether we can somehow create a fake factory method, but with the current way this was the most simple way to do it.

File: processor/src/main/java/org/mapstruct/ap/internal/model/IterableCreation.java (hunk 1 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/IterableCreation.java]: I copied this from [`SetterWrapperForCollectionsAndMapsWithNullCheck`](https://github.com/mapstruct/mapstruct/blob/871353fccbf8ed05cfc7dee055f99eb81b0e69e0/processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapperForCollectionsAndMapsWithNullCheck.java#L70-L72).

We could do it like your suggestion as well, not big difference I think. I doubt that the `EnumSet` will change it's package anytime soon 😄

**Comment by sjaakd** [comment]
From the top of my head.. isn't there a list of interfaces and respective implementations in Typefactory?

Verstuurd vanaf mijn iPhone

File: processor/src/main/java/org/mapstruct/ap/internal/model/IterableCreation.java (hunk 1 above)
Comment (chris922) [reply processor/src/main/java/org/mapstruct/ap/internal/model/IterableCreation.java]: Totally fine for me

File: processor/src/main/resources/org/mapstruct/ap/internal/model/IterableCreation.ftl (hunk 2 above)
Comment (chris922) [reply processor/src/main/resources/org/mapstruct/ap/internal/model/IterableCreation.ftl]: Okay, then let us "keep it simple, stupid" as long as we don't need a more generic way suitable for similar cases :)

**Comment by filiphr** [comment]
>From the top of my head.. isn't there a list of interfaces and respective implementations in Typefactory?

Yes, there is an `implementationType` for a certain `Type`. However, the `EnumSet` does not have an implementation type since it is a class. We then do:

```
    new
        <#if resultType.implementationType??>
            <@includeModel object=resultType.implementationType/><#if ext.useSizeIfPossible?? && ext.useSizeIfPossible && canUseSize>( <@sizeForCreation /> )<#else>()</#if>
        <#else>
            <@includeModel object=resultType/>()</#if>
        </#if>
```

**Comment by filiphr** [comment]
@chris922 if you are OK with the PR I will merge it. No need to wait for @sjaakd I think :)

**Comment by sjaakd** [comment]
No. No need 😉. 

Verstuurd vanaf mijn iPhone

**Comment by filiphr** [comment]
Thanks for the review @chris922

**Pull request merged**